LibJS: RAII in js????

davidot opened the pull request:
Still need to make some issues in the repo about spec problems, also the spec is only partly accepted so no async parts.

konradekk commented:
as much as I like the code and effort, i guess that dealing with stage 2 might not be a best idea. (there was e.g. a split with stage 2 implementation of decorators being implemented once…).

iʼm not really against it but i would advise to wait for at least stage 3 proposal.

davidot commented:
> as much as I like the code and effort, i guess that dealing with stage 2 might not be a best idea. (there was e.g. a split with stage 2 implementation of decorators being implemented once…).

This proposal is already at stage 3, although it's not mentioned in the repo itself, but it is listed here: https://github.com/tc39/proposals under stage 3 and https://github.com/babel/proposals/issues/85#issuecomment-1333747094 this is where I got the fact that async is not stage 3.

I think some stuff will still change (I even found some spec issues) but I'm up for updating the code as the proposal gets updated.

alimpfard commented on Userland/Libraries/LibJS/Runtime/AbstractOperations.cpp:
```diff
@@ -205,8 +205,8 @@ ThrowCompletionOr<void> initialize_bound_name(VM& vm, FlyString const& name, Val
     if (environment) {
-        // a. Perform ! environment.InitializeBinding(name, value).
-        MUST(environment->initialize_binding(vm, name, value));
+        // a. Perform ! environment.InitializeBinding(name, value, normal).
```
Maybe add a link to the proposal in L203 here and in the ones affected below too?

alimpfard commented on Userland/Libraries/LibJS/AST.cpp:
```diff
@@ -1096,6 +1168,11 @@ Completion ForInStatement::loop_evaluation(Interpreter& interpreter, Vector<FlyS
         // l. Let result be the result of evaluating stmt.
         auto result = m_body->execute(interpreter);
 
+        if (vm.running_execution_context().lexical_environment != old_environment) {
```
No spec comments?

alimpfard commented on Userland/Libraries/LibJS/Tests/using-for-loops.js:
```diff
@@ -0,0 +1,239 @@
+
+    test("using using of", () => {
+        let enteredLoop = false;
+        for (using using of [null]) {
```
...wtf

alimpfard reviewed (commented):
The bytecode impl is gonna be hell isn't it...

Just some random comments, nothing groundbreaking here:

davidot commented:
Addressed the comments and rebased,
also me: let's take a small break
serenity: 
![Screenshot from 2023-01-16 22-28-25](https://user-images.githubusercontent.com/3750649/212770023-37a40d76-34a6-45c7-a11d-a2bd2c2dfcd2.png)

linusg commented on Userland/Libraries/LibJS/Runtime/AbstractOperations.cpp:
```diff
@@ -1316,4 +1317,159 @@ ThrowCompletionOr<String> get_substitution(VM& vm, Utf16View const& matched, Utf
+}
+
+// 2.1.6 DisposeResources ( disposable, completion ), https://tc39.es/proposal-explicit-resource-management/#sec-disposeresources-disposable-completion-errors
+Completion dispose_resources(VM& vm, Vector<DisposableResource> const* disposable, Completion completion)
```
Can this be an `Optional<Vector<DisposableResource>> const&` instead?

davidot replied:
I've made this just a plain const reference, is that fine as well? (Turns out we cannot have nullptrs here)

linusg replied:
sure, `spec_bugs++`

davidot replied:
Well not sure it is a spec bug since they use the very dynamic types of the spec so the other one (DeclerativeEnvironments) can be null

linusg replied:
ah I see, fair enough :)

linusg commented on Userland/Libraries/LibJS/Runtime/AbstractOperations.cpp:
```diff
@@ -1316,4 +1317,159 @@ ThrowCompletionOr<String> get_substitution(VM& vm, Utf16View const& matched, Utf
+    return completion;
+}
+
+Completion dispose_resources(VM& vm, DeclarativeEnvironment const* disposable, Completion completion)
```
Likewise re ptr -> optional

davidot replied:
Environments are generally used with pointers so wrapping them in optionals is not super nice at every callsite.

Just changing it to take a `Optional<DeclarativeEnvironment const&>` means that at every callsite where we have a pointer (all of them) we have to do:
```cpp
// At this point we should have already checked that if env is not 
//  null it is a declarative environment.
Optional<DeclarativeEnvironment const&> disposable;
if (env)
	env = *static_cast<DeclarativeEnvironment>(disposable);

result = dispose_resources(vm, disposable, result);
```

As I see it we have 3 options:
- Leave it as is (not ideal)
- Add pointer constructor or something to Optional to make this nice (I don't know the implications of this but just testing it seems fine)
- Add a bunch of optional noise at every call site (I don't like this option but it's fine)

Of course if I'm missing some easy way to create the optional do let me know! I could be missing something.

linusg replied:
Alright, but can we at least upgrade those raw pointers to (nullable) `GCPtr<T>`? :)

davidot replied:
Of course!

linusg commented on Userland/Libraries/LibJS/Runtime/DisposableStack.h:
```diff
@@ -0,0 +1,40 @@
+        Disposed
+    };
+
+    [[nodiscard]] DisposableState disposable_state() const { return m_state; }
+    Vector<DisposableResource> const& disposable_resource_stack() const { return m_disposable_resource_stack; }
+    Vector<DisposableResource>& disposable_resource_stack() { return m_disposable_resource_stack; }
```
I think all new getters should be nodiscard (there might be some in earlier commits that I missed)

linusg commented on Userland/Libraries/LibJS/Runtime/DisposableStackConstructor.cpp:
```diff
@@ -0,0 +1,50 @@
+ThrowCompletionOr<Value> DisposableStackConstructor::call()
+{
+    auto& vm = this->vm();
+    return vm.throw_completion<TypeError>(ErrorType::ConstructorWithoutNew, vm.names.DisposableStack);
```
```
// 1. If NewTarget is undefined, throw a TypeError exception.
```

linusg commented on Userland/Libraries/LibJS/Runtime/DisposableStackConstructor.cpp:
```diff
@@ -0,0 +1,50 @@
+{
+    auto& vm = this->vm();
+
+    // NOTE: Step 1 is implemented in DisposableStackConstructor::call()
```
...and then this can be omitted

linusg commented on Userland/Libraries/LibJS/Runtime/DisposableStackPrototype.cpp:
```diff
@@ -0,0 +1,204 @@
+    if (disposable_stack->disposable_state() == DisposableStack::DisposableState::Disposed)
+        return vm.throw_completion<ReferenceError>(ErrorType::DisposableStackAlreadyDisposed);
+
+    auto value = vm.argument(0);
```
let's put this before step 1 for readability, it's only not needed in an error case anyway

linusg commented on Userland/Libraries/LibJS/Runtime/DisposableStackPrototype.cpp:
```diff
@@ -0,0 +1,204 @@
+    if (disposable_stack->disposable_state() == DisposableStack::DisposableState::Disposed)
+        return vm.throw_completion<ReferenceError>(ErrorType::DisposableStackAlreadyDisposed);
+
+    auto on_dispose = vm.argument(1);
```
likewise

linusg commented on Userland/Libraries/LibJS/Runtime/DisposableStackPrototype.cpp:
```diff
@@ -0,0 +1,204 @@
+    if (!on_dispose.is_function())
+        return vm.throw_completion<TypeError>(ErrorType::NotAFunction, on_dispose.to_string_without_side_effects());
+
+    auto value = vm.argument(0);
```
likewise

linusg commented on Userland/Libraries/LibJS/Runtime/DisposableStackPrototype.cpp:
```diff
@@ -0,0 +1,204 @@
+    if (disposable_stack->disposable_state() == DisposableStack::DisposableState::Disposed)
+        return vm.throw_completion<ReferenceError>(ErrorType::DisposableStackAlreadyDisposed);
+
+    auto on_dispose = vm.argument(0);
```
likewise

linusg reviewed (commented):
Great stuff! Once again we have better test coverage than test262 :smile:

davidot commented on .prettierignore:
```diff
@@ -2,3 +2,7 @@ Base/home/anon/Source/js
 Userland/Libraries/LibJS/Tests/unicode-identifier-escape.js
 Userland/Libraries/LibJS/Tests/modules/failing.mjs
+
+# FIXME: Remove once prettier is updated to support using declarations.
```
So should I just add it as a comment? I mostly left it as a FIXME to make it easier to find for random fixme things (hopefully once it is stage 4)

linusg replied:
nah, FIXME is fine, just thinking out loud :)

davidot commented on Userland/Libraries/LibJS/Runtime/AbstractOperations.cpp:
```diff
@@ -1316,4 +1317,159 @@ ThrowCompletionOr<String> get_substitution(VM& vm, Utf16View const& matched, Utf
+            return vm.throw_completion<TypeError>(ErrorType::NoDisposeMethod, value.to_string_without_side_effects());
+    }
+    // 2. Else,
+    // a. If IsCallable(method) is false, throw a TypeError exception.
+    // NOTE: This is guaranteed to never occur from the type.
```
I'll put it on my list!

davidot commented:
Addressed most of the comments, just not sure what would be best in the Optional reference case

linusg commented on Userland/Libraries/LibJS/Parser.cpp:
```diff
@@ -3050,6 +3076,55 @@ NonnullRefPtr<VariableDeclaration> Parser::parse_variable_declaration(bool for_l
     return declaration;
 }
 
+NonnullRefPtr<UsingDeclaration> Parser::parse_using_declaration(JS::Parser::IsForLoopVariableDeclaration is_for_loop_variable_declaration)
```
We're already in the parser here, no need to specify `JS::Parser` (also in two other places)

linusg replied:
or maybe `Parser::` is needed? But definitely not `JS::` :)

davidot replied:
That was just me trusting the CLion auto complete without checking

linusg commented on Userland/Libraries/LibJS/Runtime/DisposableStackConstructor.cpp:
```diff
@@ -0,0 +1,49 @@
+// 11.3.1.1 DisposableStack ( ), https://tc39.es/proposal-explicit-resource-management/#sec-disposablestack
+ThrowCompletionOr<Value> DisposableStackConstructor::call()
+{
+    // 1. If NewTarget is undefined, throw a TypeError exception.
```
nit: should go after `vm`

davidot commented:
Done and also quickly checked all the commit do build and run `test-js` since I had some refactor trouble.

linusg reviewed (approved):
Thanks for the initial implementation, great work! :tada:
